p2p/server/deadline_adjuster.go

```diff
@@ -0,0 +1,130 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/libp2p/go-yamux/v4"
+)
+
+const (
+	deadlineAdjusterChunkSize = 4096
+)
+
+type deadlineAdjuster struct {
+	peerStream
+	desc            string
+	timeout         time.Duration
+	hardTimeout     time.Duration
+	totalRead       int
+	totalWritten    int
+	start           time.Time
+	clock           clockwork.Clock
+	chunkSize       int
+	nextAdjustRead  int
+	nextAdjustWrite int
+	hardDeadline    time.Time
+}
+
+func newDeadlineAdjuster(stream peerStream, desc string, timeout, hardTimeout time.Duration) *deadlineAdjuster {
+	return &deadlineAdjuster{
+		peerStream:      stream,
+		desc:            desc,
+		timeout:         timeout,
+		hardTimeout:     hardTimeout,
+		start:           time.Now(),
+		clock:           clockwork.NewRealClock(),
+		chunkSize:       deadlineAdjusterChunkSize,
+		nextAdjustRead:  -1,
+		nextAdjustWrite: -1,
+	}
+}
+
+func (dadj *deadlineAdjuster) augmentError(what string, err error) error {
+	if !errors.Is(err, context.DeadlineExceeded) && !errors.Is(err, yamux.ErrTimeout) {
+		return err
+	}
+	now := dadj.clock.Now()
+	return fmt.Errorf("%s: %s after %v, %d bytes read, %d bytes written, timeout %v, hard timeout %v: %w",
+		dadj.desc,
+		what,
+		now.Sub(dadj.start),
+		dadj.totalRead,
+		dadj.totalWritten,
+		dadj.timeout,
+		dadj.hardTimeout,
+		err)
+}
+
+func (dadj *deadlineAdjuster) adjust() error {
+	now := dadj.clock.Now()
+	if dadj.hardDeadline.IsZero() {
+		dadj.hardDeadline = now.Add(dadj.hardTimeout)
+	} else if !now.Before(dadj.hardDeadline) {
+		// emulate yamux timeout error
+		return yamux.ErrTimeout
+	}
+	// Do not adjust the deadline too often
+	adj := false
+	if dadj.totalRead > dadj.nextAdjustRead {
+		dadj.nextAdjustRead = dadj.totalRead + dadj.chunkSize
+		adj = true
+	}
+	if dadj.totalWritten > dadj.nextAdjustWrite {
+		dadj.nextAdjustWrite = dadj.totalWritten + dadj.chunkSize
+		adj = true
+	}
+	if adj {
+		// We ignore the error returned by SetDeadline b/c the call
+		// doesn't work for mock hosts
+		deadline := now.Add(dadj.timeout)
+		if deadline.After(dadj.hardDeadline) {
+			_ = dadj.SetDeadline(dadj.hardDeadline)
+		} else {
+			_ = dadj.SetDeadline(deadline)
+		}
+	}
+
+	return nil
+}
+
+func (dadj *deadlineAdjuster) Read(p []byte) (n int, err error) {
+	var nCur int
+	for n < len(p) {
+		if err := dadj.adjust(); err != nil {
+			return n, dadj.augmentError("read", err)
+		}
+		to := min(len(p), n+dadj.chunkSize)
+		nCur, err = dadj.peerStream.Read(p[n:to])
+		n += nCur
+		dadj.totalRead += nCur
+		if err != nil {
+			return n, dadj.augmentError("read", err)
+		}
+		if n < to {
```

Comment:
shouldn't this be:
```suggestion
		if nCur < to {
```

Reply:
No, `nCur` is the count for the current read only, whereas `n` and `to` are positions in the buffer (target slice).
If the requested number of bytes is read then after
```go
		n += nCur
```
(slightly above this part) `n` should equal `to`. Otherwise, it's a short read.

Reply:
Hm, then this condition could only be satisfied on the first loop iteration. Once `n >=to` it will stay this way as `n` only increases and `to` only decreases (or stays the same). Is it intentional? I thought that this loop should break when `Read()` returns less data than asked for (`to`).

Reply:
Let's look at some loop iterations. Let's say we want to read 10000 bytes (that's `len(p)`) and chunk size is 4096

```
Step 1, before Read:
  n=0 to=4096
  after Read() of 4096 bytes (nCur == 4096) and increment:
  n=4096 to=4096
Step 2, before Read:
  n=4096 to=8192
  after Read with nCur==4096 and increment:
  n=8192 to=8192
Step 3, before Read:
  n=4096 to=10000
  after Read with nCur==1808 and increment:
  n=10000 to=10000 => loop terminates b/c n == len(p)
```

Now let's say on step 2 we get a short read:

```
Step 1, before Read:
  n=0 to=4096
  after Read() of 4096 bytes (nCur == 4096) and increment:
  n=4096 to=4096
Step 2, before Read:
  n=4096 to=8192
  after Read with nCur==2000 and increment:
  n=6096 to=8192 ==> loop terminates b/c 6096 < 8192 (n < to)
```

Reply:
Thanks, I missed the fact that `to` increases as the reading progresses.